Contest base code

osherwott opened the pull request:
Implements Contests as a battle mechanic, along with a contest GameState.
~~Contests are far from finished, but this is a start.~~ Well that was a lie, contests are now basically ready. Only things left are pokeblocks, requirements, and maybe some adjustment on CT rewards.

## Description
### Contest types
Added contest types to pokemon based off their highest base stats, inspired from natures and flavors of the games.
- `Cool` - Attack - `Spicy`
- `Beautiful` - Special Attack - `Dry`
- `Cute` - Speed - `Sweet`
- `Smart` - Special Defense - `Bitter`
- `Tough` - Defense - `Sour`
If a pokemon has multiple highest stats it has multiple Contest Types.

Contests types are also additive, and offer a multiplier of `0.5`, `1`, or `2`.

If a pokemon already has a contest type given to it in the PokemonList, it will keep that over its assumed type. This has been done to all contest shop pokemon, and some other forms for fun (ex. a "Balanced" Surfing Pikachu, Pinkans, who eat sweet berries, etc.). I'd like for most forms to have different contest types and will likely go back to any skipped ones later.

### Appeal
Added appeal as an attack equivalent for contests. These can be increased by pokeblocks and have a given a value of 1 for now. The 5 basic blocks can be fed to a pokemon of that block's flavor/contest type association and all blocks can be used on `Balanced` pokemon.

Contest types, appeal, and pokeblocks can be viewed in the pokemon statistics modal after getting the Pokeblock case key item. Until pokeblocks are implemented, the Pokeblock case and blocks can be found for testers in Lilycove's contest shop.

### Contest feature
Contests are split into ranks and contest types. Each rank is supposed to scale in difficulty. Damage is based on appeal and contest types. The locations of each rank is the same as the contest halls in Ruby and Sapphire. This decision was made to give empty towns more content.
Contest Halls are town content that organize the Contests available by rank and type.
#### Ranks in order (with locations):
- `Normal` in Verdanturf
- `Super` in Fallarbor
- `Hyper` in Slateport
- `Master` in Lilycove
- `Super Normal`, `Super Great`, `Super Ultra`, and `Super Master` in Hearthome
- `Spectacular` in Lilycove
- `Brilliant Shining` in Hearthome
Each rank has an appeal number associated with it that influence each rank. These were determined by referencing the difference between each rank in the games and applying it to one large continuum. The values can be found in `rankAppeal` in `ContestHelper.ts` and are the same as the last three `multiplier` options in `PokeBlockController.ts`: `80`, `150`, and `220`.

#### Contest Battles
Contests have two "health" bars, one is for the enemy, and one is for the contest's audience. Enemies are incremented through clicking, and audience appeal is incremented each tick. Enemies have varying contest types but the audience's depends on the type of contest the player chooses to enter in the Contest Hall. Audience and Enemy HP increases with each bonus round.

#### Winning, Contest Tokens, and Bonus "encore" Rounds
A contest is won when the audience bar is maxed when the time is up. A base amount of Contest Tokens are rewarded based on the rank, with a bonus depending on the number of trainers defeated at the end of the round. If the audience bar is filled out before or at 3 seconds left, a bonus round will be activated, letting players continue their trainer streak and earn more Tokens. Each bonus round resets the audience bar and increases the amount of max appeal needed to fill it.  If a bonus round is lost, only a few bonus tokens are awarded. The max amount of bonus rounds is dependent on the rank. Tokens are rewarded upon finishing each round, so they are free to leave whenever they please.

Since contests are mostly a battle with time, the time flute is available to be activated in them.

### "Rally" feature - opposite of Damage
Added a new way to "defeat" pokemon by increasing the "health" bar. Only contest audiences use it now, but it was added to BattlePokemon for future content, like rallying wild encounters for a possible ranger region, or for party pokemon and some other mechanic.

## Motivation and Context
- Adds contest types and an appeal stat to pokemon to bring some depth to regular gameplay
- Related to #5105, a way to make contests more dynamic

## How Has This Been Tested?
- Tested damage received based on contest types and the TypeHelper for them
- Tested winning and losing a contest and its bonus rounds
- Tested entering and leaving a contest and contest GameState
- Tested increasing pokemon appeal with pokeblocks

### Other
Anything past the first four Hoenn contests are not finalized and only exist for feedback on the general difficulty curve.

## Screenshots:
Battle View
![image](https://github.com/pokeclicker/pokeclicker/assets/114451054/02dd514d-3f2c-45ef-a2a2-50f1283d3dde)
Pokeblock modal
![image](https://github.com/pokeclicker/pokeclicker/assets/114451054/f3f9975e-9e72-4fd0-94db-67564cee75a5)

## Types of changes
- New feature

osherwott commented:
Thanks for the feedback, those are all good ideas!

> not adding contest code to `BattlePokemon`

Any ideas on approaching this? Pokemon need to have their contest types, and I'm unsure how to do this otherwise.

> `Party.calculatePokemonAttack()`

This was to make contest code more relevant outside of its main mechanic, but I agree that it's something that can be figured out later. I will revert those changes.

> everything else

Should be easy to do 👍 

> Are these contests going to be entirely separate from PokemonContest? Progress on that seems to have stalled for a while.

Most likely. I plan to phase out the old code, but not completely. It'd be nice if we could salvage the mechanic somehow, but it's not very intuitive for what contests have the potential to be. I think progress is stalled on it because everyone knows I've been working on these instead...

osherwott commented on src/scripts/party/PartyPokemon.ts:
```diff
@@ -712,6 +808,26 @@ class PartyPokemon implements Saveable {
+
+    set currentContestType(currentContestType: ContestType[]) {
+        this._currentContestType(currentContestType);
+    }
```
I cleaned up according to your other suggestions, but do you think this is still something I should change? There shouldn't be any pokemon, battle or party, that has the None type now.

osherwott replied:
I resolved this by removing the `None` contest type

umbralOptimatum reviewed (changes requested):
Noticed these when looking at your changes just now. Apologies for not catching these; I was focused on PartyPokemon at the time.

osherwott commented on src/modules/types/ContestTypeHelper.ts:
```diff
@@ -0,0 +1,71 @@
+        // eslint-disable-next-line no-param-reassign
+        d3 = d3 !== ContestType.None ? d3 : 6;
+
+        let m1 = ContestTypeHelper.contestTypeMatrix[a1][d1];
+        let m2 = ContestTypeHelper.contestTypeMatrix[a2][d1];
+        let m3 = ContestTypeHelper.contestTypeMatrix[a3][d1];
+        
+        let m4 = ContestTypeHelper.contestTypeMatrix[a1][d2];
+        let m5 = ContestTypeHelper.contestTypeMatrix[a2][d2];
+        let m6 = ContestTypeHelper.contestTypeMatrix[a3][d2];
+
+        let m7 = ContestTypeHelper.contestTypeMatrix[a1][d3];
+        let m8 = ContestTypeHelper.contestTypeMatrix[a2][d3];
+        let m9 = ContestTypeHelper.contestTypeMatrix[a3][d3];
+
+        // Find the numbers for each type, cap at 1
+        let type1 = Math.min(m1 + m2 + m3, 1);
+        let type2 = Math.min(m4 + m5 + m6, 1);
+        let type3 = Math.min(m7 + m8 + m9, 1);
+
+        // Find the most effective type and use it
+        return Math.max(type1, type2, type3);
```
Type balancing was a bit troublesome. The intention is to be different from TypeHelper multipliers.

I approached this with the idea of audience appeal in mind, so if a pokemon is able to "max" appeal, I want that to be registered over an immunity. Ex. If a pokemon in a Cool contest is Cool and Cute, the audience won't dislike it for being Cute, because they only care if it's being Cool, therefore the closer it is to Cool (which is where adjacent types like Tough and Beautiful in this case come in), the more "points" it gets. I dabbled with the idea of different calculations for audience and opponent pokemon, but that just made everything too messy, so I prioritized the audience part.

The three type sweet spot is also intentional, although I'm a bit less certain on whether it's the right move. The idea here is to give incentive to feed type-adding pokeblocks to pokemon. Only a handful of pokemon have naturally-occurring triple types... but there's nothing stopping you from putting all your regular blocks in just them. I have always envisioned ribbons to help encourage feeding multiple pokemon, but it would be preferable to have a gameplay element do that too somehow. Of course, it would also be possible to redo the types completely, and even remove `Balanced`, if you're up to some extreme modifications. I kind of felt my way through with the current type chart iteration, so I wouldn't be surprised if there's something better out there.

osherwott replied:
This is all great feedback and I'm really grateful you're getting into the nitty gritty for me. Game mechanics are always a tough one. Let me lay down my intentions with the ranks and bars first, both to get feedback from you, and because I think better by writing out my thoughts.
Contests are divided into four sets, each with their own gimmick:
- Hoenn's `Normal` - `Master`: no gimmick, introductory
- Sinnoh's `Super Normal` - `Super Master`: some gimmick
- Hoenn's `Spectacular`: different gimmick from regular Sinnoh set
- Sinnoh's `Brilliant Shining`: have regular Sinnoh and `Spectacular` gimmicks come together

Within each contest itself, we have two things going on:
- Audience: this determines if you've won the contest or not
  - These take on the contest type of the contest that is selected, so the points dealt to the bar in each contest is different
- Trainers: this is meant to be the main way of gaining Contest Tokens, hence the Trainer Streak

So we have, in theory, a way to influence damage by rank and contest type, and we are left with audience and trainer bars that are affected by a similar formula. You mentioned some good ideas for what I'm assuming is the trainer bar? Here:
> Maybe if the opponent's pokemon matches the contest type, it always defends with that type and gets a bonus, but if not your pokemon get to attack its weakest type.

You'll have to elaborate on what you mean by "always defends with that type", but I like the idea of a bonus and further connecting the opponent with the contest's `ContestType`. If it's done well enough I can even scrap the bonus upon defeating it.

You also mentioned a battlefield effect. `Spectacular` rank's type-restricting gimmick is the closest to that, but is too extreme to apply across all ranks, and it still has the problem of the audience bar increasing with the same formula as the trainer bar. This is where the idea of Sheen comes into play. Sheen could influence how much attack is being dealt to the opponent. There could be a debuff affecting each rank's opponents, and while a pokemon who's type matches the contest's wouldn't need sheen to attack the opponent with 100% of its appeal, other pokemon would need to offset the debuff with a high enough sheen. I'm unsure if this overcomplicates things, and also if it should exist for the introductory Hoenn contests or not, but it's something I just came up with and curious to see if it would lead to somewhere.

osherwott commented on src/modules/types/ContestTypeHelper.ts:
```diff
@@ -0,0 +1,71 @@
+        ];
+    })();
+
+    public static getAppealModifier(
+        a1: ContestType = ContestType.None,
+        a2: ContestType = ContestType.None,
+        a3: ContestType = ContestType.None,
+        d1: ContestType = ContestType.None,
+        d2: ContestType = ContestType.None,
+        d3: ContestType = ContestType.None,
+    ): number {
```
Just got a good look at this and wow it's actually so nice and concise. It's making me rethink the type distribution just a little bit. If we use arrays then all the `Balanced` pokemon can actually have their 4 or 5 default contest types, and `Balanced` can be reserved for really special things like the high ranking contests and alt forms.

osherwott committed:
Merge branch 'develop' into contest-appeal-attack

osherwott commented on src/scripts/contest/ContestHelper.ts:
```diff
@@ -0,0 +1,130 @@
+        if (includeBreeding || !pokemon.breeding) {
+            appeal = pAppeal * ContestTypeHelper.getAppealModifier(pType1, pType2, pType3, type1, type2, type3);
+        }
+
```
Arrays worked like a charm! Everything looks so clean now~ Thanks!

osherwott committed:
contest types styling

huh never committed this for some reason

osherwott committed:
fix key item stuff for the last time

osherwott committed:
fix contest type pokeblocks

osherwott committed:
clicking for all ranks

osherwott committed:
event click bonuses for hoenn ranks

osherwott committed:
some kind of reward refactor

osherwott committed:
if undefined contesttypehelper return 0

osherwott committed:
Contest stats and requirement

osherwott committed:
change appeal calculations

now it increments depending on the pokemons rank bracket

osherwott committed:
increment contestroundswon

osherwott committed:
enemy array, multiple opponents

osherwott committed:
remove click types and combos

that is basically tracked by trainers defeated now

osherwott committed:
change buttons so contest doesn't start right away

osherwott committed:
another special event trainer

osherwott committed:
Update .eslintrc.js

osherwott committed:
contestWonModal

osherwott committed:
couple more requirements

osherwott committed:
altarianite

osherwott committed:
stylings

osherwott committed:
contest trainer rewards for endless rank

osherwott committed:
everything else

osherwott committed:
sheen pokeblock for testing

osherwott committed:
remove test sinnoh hall

osherwott committed:
Merge branch 'develop' into contest-appeal-attack

osherwott committed:
lint

osherwott committed:
Merge branch 'contest-appeal-attack' of https://github.com/osherwott/pokeclicker into contest-appeal-attack

osherwott committed:
Update contest.less

osherwott committed:
fix appeal function

osherwott committed:
fix appeal calc again

we'll play with sheen some other time

osherwott committed:
Update ContestHelper.ts

osherwott committed:
min of 0.1 appeal

osherwott committed:
Trainer info for testing

osherwott committed:
clean up generateContestPokemon and add gender

- don't need exp
- add gender for consistent sprites

osherwott committed:
Remove sinnoh contests

osherwott committed:
some cleanup

osherwott committed:
fix bug for Spectacular crowdHype

osherwott committed:
finalize rewards

- defeating a trainer stores rewards
- rewards only given if contest is won

osherwott committed:
more testing info

osherwott committed:
better click appeal

- base click increases for every contest that has been cleared once
- use gold pokeblocks for boost
- allows no click challenges to get Spectacular rewards too

osherwott committed:
couple more statistics

since people seem to love em

osherwott committed:
linting

osherwott committed:
lil fixes

- gold pokeblocl click appeal tooltip
- sentence about increasing click appeal every contest cleared
- fix idle `defeatPokemon`

osherwott committed:
mobile support

osherwott committed:
Prevent Contest Start if 0 appeal, notif. titles

osherwott committed:
give exp back

osherwott committed:
base click of 1

osherwott committed:
Update contest.less

osherwott committed:
Update TownList.ts

osherwott committed:
Update BattlePokemon.ts

osherwott committed:
Update Game.ts

osherwott committed:
Update GameController.ts

osherwott committed:
Update PartyController.ts

osherwott committed:
Create ContestScore.ts

osherwott committed:
Update ContestHall.ts

osherwott committed:
Update Egg.ts

osherwott committed:
Update ContestTypeHelper.ts

osherwott committed:
Update temporaryWindowInjection.ts

osherwott committed:
Update index.ts

osherwott committed:
Create InContestTypeRequirement.ts

osherwott committed:
Update InContestRequirement.ts

osherwott committed:
Create InContestRankRequirement.ts

osherwott committed:
Update ContestWonRequirement.ts

osherwott committed:
Update BerryUnlockedRequirement.ts

osherwott committed:
Update temporaryWindowInjection.ts

osherwott committed:
Create Direction.ts

osherwott committed:
Create ContestOpponentStatus.ts

osherwott committed:
Update index.ts

osherwott committed:
Create ContestBattleSpectacular.ts

osherwott committed:
Update ContestBattlePokemon.ts

osherwott committed:
Create ContestBattleDefault.ts

osherwott committed:
Update ContestBattleSpectacular.ts

osherwott committed:
Create ContestBattleDance.ts

osherwott committed:
Update ContestPokemon.ts

osherwott committed:
Update PokemonFactory.ts

osherwott committed:
array

osherwott committed:
Update ContestBattleDefault.ts

osherwott committed:
Update ContestBattleDance.ts

osherwott committed:
Update ContestBattleDance.ts

osherwott committed:
Create ContestBattleDancePose.ts

osherwott committed:
poses -> dance combos

osherwott committed:
Update ContestBattleDance.ts

osherwott committed:
Update contest.less

osherwott committed:
Update ContestTrainer.ts

osherwott committed:
Update ContestBattleDance.ts

osherwott committed:
Update ContestTrainer.ts

osherwott committed:
Update ContestRunner.ts

osherwott committed:
Update ContestHelper.ts

osherwott committed:
Update ContestHall.ts

osherwott committed:
Update ContestBattle.ts

osherwott committed:
Update badgeCase.html

osherwott committed:
Update pokemonStatisticsModal.html

osherwott committed:
Create contestSpotlight.png

osherwott committed:
Update ContestHelper.ts

osherwott committed:
pushing for testing to work, will review later

osherwott committed:
Update PokeBlock.ts

osherwott committed:
Update ItemNameType.ts

osherwott committed:
Update ItemList.ts

osherwott committed:
Update GameConstants.ts

osherwott committed:
Update pokeblockModal.html

osherwott committed:
trainer sprites

osherwott committed:
Update itemModal.html

osherwott committed:
Update contestWonModal.html

osherwott committed:
Update contestView.html

osherwott committed:
Update contestInfo.html

osherwott committed:
ribbon badge case rows

osherwott committed:
Update contestWonModal.html

osherwott committed:
oh

osherwott committed:
Update contestInfo.html

osherwott committed:
merging

osherwott committed:
images

osherwott committed:
other notes

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
Update HotkeySetting.ts

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
Merge remote-tracking branch 'upstream/develop' into contest-appeal-attack

osherwott committed:
Update translations

osherwott committed:
Update ContestHall.ts

osherwott committed:
lint

osherwott committed:
lint

osherwott committed:
not lint

osherwott committed:
Update PokemonFactory.ts

osherwott committed:
contest trainer modules

osherwott committed:
contest score module

osherwott committed:
ribbon svg module

osherwott committed:
Delete InContestRequirement.ts

osherwott committed:
contest combo module

osherwott committed:
remove in global

osherwott committed:
interfaces

osherwott committed:
sinnoh hall in hoenn for easy testing

osherwott committed:
contest pokemon module

osherwott committed:
contest battle pokemon module

osherwott committed:
spectacular battle module

osherwott committed:
default battle module

osherwott committed:
dance contest module

osherwott committed:
Create ContestRewards.ts

osherwott committed:
import to requirement modules

osherwott committed:
Update contestWonModal.html

osherwott committed:
Update PokemonFactory.ts

osherwott committed:
main modules

osherwott committed:
Update ContestHall.ts

osherwott committed:
temp window stuff

osherwott committed:
update html

osherwott committed:
Update buyKeyItem.ts

osherwott committed:
encore bonus math

osherwott committed:
remove random notes

osherwott committed:
Update getContestRankTypeProxy.ts

osherwott committed:
Update ContestBattleDance.ts

osherwott committed:
Update ContestBattle.ts

osherwott committed:
unify rewards, compensate berries in dance halls

osherwott committed:
eslint and stuff

osherwott committed:
better berry compensation logic

osherwott committed:
better scoring logic, clear up some info

osherwott committed:
Update badgeCase.html

osherwott committed:
Update ContestRunner.ts

osherwott committed:
timers and dancing tweaks

osherwott committed:
hide item rewards for testing

osherwott committed:
Update contestInfo.html

osherwott committed:
hoenn test hall

osherwott committed:
Update ContestHelper.ts

osherwott committed:
Update ContestRunner.ts

osherwott committed:
ui

osherwott committed:
tweak dancing rewards

osherwott committed:
Update TownList.ts

osherwott committed:
Update TownList.ts

osherwott committed:
Update PartyPokemon.ts

osherwott committed:
Update EvolutionHandler.ts

osherwott committed:
Update TemporaryScriptTypes.ts

osherwott committed:
Update ItemList.ts

osherwott committed:
Update GameConstants.ts

osherwott committed:
Update ContestRunner.ts

osherwott committed:
html

osherwott committed:
Update ContestHelper.ts

osherwott committed:
Update ContestHelper.ts

osherwott committed:
Update ContestHelper.ts

osherwott committed:
Revert "Update TownList.ts"

This reverts commit 41880396368011612e68851a04088ee84c1a89c7.

osherwott committed:
white pokeblock desc

osherwott committed:
Update ItemList.ts

osherwott committed:
better reduceSheen function

osherwott committed:
fixes

osherwott committed:
testing toggle

osherwott committed:
Update contestView.html

osherwott committed:
Update ItemList.ts

osherwott committed:
Update TownList.ts

osherwott committed:
Update ContestHelper.ts

osherwott committed:
Update TownList.ts

osherwott committed:
pikachus

osherwott committed:
misc

osherwott committed:
Update TownList.ts

osherwott committed:
shopping

note to self: pokeblocks are mostly a hint to what type is being sold. would be nice to have pikas need a unique drop from the breeder trainer but maybe some other time

osherwott committed:
Update ContestTrainerList.ts

osherwott committed:
Update GenericDeal.ts

osherwott committed:
better

osherwott committed:
Update ContestHelper.ts

osherwott committed:
calcs

osherwott committed:
more contest shops and tweaks

osherwott committed:
numbers x10 for no decimals

osherwott committed:
misc

osherwott committed:
Update PokeBlockController.ts

osherwott committed:
fixes

osherwott committed:
Update ContestBattle.ts

osherwott committed:
Merge branch 'develop' into contest-appeal-attack

osherwott committed:
Update ContestRunner.ts

osherwott committed:
scarves

osherwott committed:
better scarves html

osherwott committed:
fixes

osherwott committed:
Update index.ts

osherwott committed:
Update PokemonList.ts

osherwott committed:
pokeblocks image revisit

osherwott committed:
uniform pokeblock effects

osherwott committed:
Update ItemList.ts

osherwott committed:
Ranks, Types enums, Statistics, and TypeHelper

basic building block enums
type chart is additive, caps at 1
statistic proxy taken from route

osherwott committed:
Contest Battle Pokemon

all this will be used with Contest Trainers to generate opponents in PokemonFactory
introduces `rapport` as an opposite to health (fills the bar up instead of down), can be used in conjunction to health, but it's used by itself here
3 basic opponent statuses, more to come later

osherwott committed:
Contest Trainers, generate opponents

trainer options only have `requirement` for now, will include more later
new trainer images will also be added later

osherwott committed:
Basic controls, Game ticks

osherwott committed:
Create ContestScore.ts

one of the main gameplay files
this will be used for scoring and determining contest token gain, as well as other things

osherwott committed:
Create ContestHelper.ts

place for misc functions

osherwott committed:
Main Gameplay code

osherwott committed:
Update temporaryWindowInjection.ts

osherwott committed:
Contest Hall and Requirements

osherwott committed:
Contest Won modal

mostly a placeholder

osherwott committed:
Info modal

pop-up modal for general info as well as rank-specific gameplay, trainers, and rewards

osherwott committed:
Contest Control Panel

osherwott committed:
contestView.html header

Includes title and timer
Title also shows completion status
I don't like the horizontal scaling function but I couldn't get a dynamic font size with anything i tried. I don't want to add exceptions for all the rank, type, and theme combinations, and I don't want every title to be small just to fit the future rank names

osherwott committed:
contestView.html start view and opponents

Start button, Trainers, and Pokemon
spotlight status will change depending on the game mode, but is simple for hoenn ranks
`healthDisplayBeat` is separate because it's used in the info modal by itself
the empty space below the Start button will be used to show warnings for not having party contest stats

osherwott committed:
contestView.html audience bar

now in technicolor!

osherwott committed:
contestView.html table

osherwott committed:
linting for one thing

osherwott committed:
Update contest.less

osherwott committed:
Add getTeam() as ContestPokemon in ContestTrainer

osherwott committed:
Simplify opponent generation, add better comments

The old line is obsolete and from earlier drafts where opponents weren't assessed in batches

osherwott committed:
Move rapport and rally to ContestBattlePokemon

Keeps BattlePokemon clean.
I still want this as a general reference for an "opposite of health" bar, which is why I've kept the parameter names generic

osherwott committed:
Merge branch 'develop' into contest-part-minigame

osherwott committed:
Merge branch 'contest-part-minigame' into contest-appeal-attack

osherwott committed:
merging

osherwott committed:
merging

osherwott committed:
Update HeldItem.ts

osherwott committed:
Update contest.less

umbralOptimatum committed:
Merge branch 'develop' into contest-part-minigame

umbralOptimatum committed:
Small cleanup

osherwott committed:
Merge branch 'contest-part-minigame' into contest-appeal-attack
